### ISSUE TITLE
Naming in `Schema` world

### ISSUE THREAD

**Issue by sandreae** [p2panda/p2panda#261]
## CDDL definitions
- `operation_format`
- cddl definition for `SchemaXyz`

### System Schema
- `SchemaDefinition`
- `SchemaFieldDefinition`
- `KeyGroup`

### Application Schema
- `Name`

## Schema IDs
- `SchemaId::SchemaDefinition` // `"schema_definition_v1"`
- `SchemaId::SchemaFieldDefinition` // `"schema_field_definition_v1"`
- `SchemaId::Application(PinnedRelation)` // `"HASH"`

Accompanying diagram here: https://whimsical.com/schemas-3e8BPkS2aAry64NFS2ynp4

@cafca and @sandreae achieved consensus on the above

**Comment by adzialocha** [comment]
Hui, thank you so much for writing all of this down here!

I think I agree with the whole general bit but have some struggles with the way it is summarized here, maybe its exactly what you also have in mind but just want to clarify.

* You mention the category "CDDL definitions" where "Operation Format" falls in, but not "SchemaDefinition", even though they are also CDDL definitions. I liked @cafca's idea to not focus too much on CDDL at all, so I would rename "CDDL definitions" to just "Operations"?
* Why are the system schemas "SchemaDefinition" and "SchemaFieldDefinition" ending with "..definition" but not KeyGroup?
* I didn't understand "Schema Definition" as the name, but rather as "The definition of the system schema 'schema', the definition of the system schema "key group'", so reading "SchemaId::SchemaDefinition" or "schema_definition_v1" is a little bit strange. How about just saying "schema_v1" when you address it in code? When you talk about it you can say "This is the schema definition v1"
* Between humans it makes sense to relate to an application schema with its name, programatically we will use the "name_${entry-hash}" to identify them (see https://github.com/p2panda/aquadoggo/blob/4f0deef606c6789b7e22d0c1e86cd3044cf12e7a/aquadoggo/src/graphql/convert/schema.rs#L28, ignore the bad name plz) as names are not unique
* What does `"HASH"` mean?

Generally speaking: The operation format and system schema definitions are specifications which define the shape of data. We express this in form of words and diagrams like "An operation has a `version` field which is an integer starting with "1", it has an action field which can be "update", "delete" etc.." and validate this definition with the help of CDDL but potentially something else some day. 

My suggestion would be like that:

## Operations

- `operation_format`

## Schemas

### System Schemas

- `SchemaDefinition`
- `SchemaFieldDefinition`
- `KeyGroupDefinition`
- ...

### Application Schemas

- `Name_${EntryHash}`

## Schema IDs

- `SchemaId::Schema` // `"schema_v1"`
- `SchemaId::SchemaField` // `"schema_field_v1"`
- `SchemaId::Application(PinnedRelation)`

**Comment by sandreae** [comment]
I like this a lot :+1: you had the same understanding of all the concepts we're describing, it's good that we at least agree on that now :rofl: I think this way of naming and grouping is good.

**Comment by adzialocha** [comment]
I see now the diagram and the category of "System Schema Formats", thats also interesting ..

**Comment by sandreae** [comment]
Yeh, I like that toooo.

**Comment by adzialocha** [comment]
> Yeh, I like that toooo.

Ah, I think in that context I understand some parts better now :+1: Sorry, just wrapping my head around .. In this case it would look like that, right?

## Specifications

- Operation format
- "Schema Definition" format
- "Schema Field Definition" format
- "Key Group" format

## Concept: Schemas

### System Schemas

- Schema definition
- Schema Field definition
- Key Group
- ...

### Application Schemas

- Name

## Implementation: Schema IDs

- `SchemaId::SchemaDefinition` // `"schema_definition_v1"`
- `SchemaId::SchemaFieldDefinition` // `"schema_field_definition_v1"`
- `SchemaId::KeyGroup`
- `SchemaId::Application(PinnedRelation)` // `"${name}_${hash}"`

**Comment by sandreae** [comment]
Yes, exactly! The "definition" suffix was added to give the specific schema "Schema" types some distinction.

**Comment by adzialocha** [comment]
Cool, I like it!

**Comment by sandreae** [comment]
We can say that `SchemaDefinition` + `[SchemaFieldDefinition]` = `Schema` (or `ApplicationSchema`)

**Comment by sandreae** [comment]
This is nice in it's language, and also correct in terms of the spec/code.

**Comment by cafca** [comment]
This seems more like a handbook issue to me and could be closed by a handbook update. Can we move it there?

**Comment by adzialocha** [comment]
> This seems more like a handbook issue to me and could be closed by a handbook update. Can we move it there?

Sounds good!
